File: packages/android/src/main/java/com/kozmos/components/Menu/Menu.kt
package com.kozmos.components.menu

import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Divider
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens

enum class KozmosMenuContentType {
    Basic,
    Checkbox,
    Radio,
    Submenu
}

data class KozmosMenuItem(
    val text: String,
    val shortcut: String? = null,
    val checked: Boolean = false,
    val selected: Boolean = false,
    val submenuItems: List<String> = emptyList()
)

data class KozmosMenuContent(
    val label: String? = null,
    val items: List<KozmosMenuItem>,
    val contentType: KozmosMenuContentType = KozmosMenuContentType.Basic
)

@Composable
fun KozmosMenu(
    expanded: Boolean,
    onDismissRequest: () -> Unit,
    items: List<String>,
    onItemClick: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    KozmosMenu(
        expanded = expanded,
        onDismissRequest = onDismissRequest,
        content = KozmosMenuContent(
            items = items.map { KozmosMenuItem(text = it) }
        ),
        modifier = modifier,
        onItemClick = { item -> onItemClick(item.text) }
    )
}

@Composable
fun KozmosMenu(
    expanded: Boolean,
    onDismissRequest: () -> Unit,
    content: KozmosMenuContent,
    modifier: Modifier = Modifier,
    onItemClick: (KozmosMenuItem) -> Unit = {}
) {
    DropdownMenu(
        expanded = expanded,
        onDismissRequest = onDismissRequest,
        modifier = modifier
    ) {
        content.label?.let { label ->
            DropdownMenuItem(
                text = {
                    Text(
                        text = label,
                        color = KozmosThemeTokens.primitivesColorsForeground500
                    )
                },
                onClick = {}
            )
            Divider(color = KozmosThemeTokens.semanticsBorderSubtle)
        }

        content.items.forEach { item ->
            DropdownMenuItem(
                text = { Text(item.text) },
                onClick = {
                    onItemClick(item)
                    onDismissRequest()
                },
                leadingIcon = {
                    when (content.contentType) {
                        KozmosMenuContentType.Checkbox ->
                            if (item.checked) {
                                Icon(Icons.Default.Check, contentDescription = null)
                            } else {
                                Spacer(modifier = Modifier.width(24.dp))
                            }
                        KozmosMenuContentType.Radio ->
                            if (item.selected) {
                                Icon(Icons.Default.Check, contentDescription = null)
                            } else {
                                Spacer(modifier = Modifier.width(24.dp))
                            }
                        else -> Unit
                    }
                },
                trailingIcon = {
                    when {
                        content.contentType == KozmosMenuContentType.Submenu && item.submenuItems.isNotEmpty() ->
                            Icon(Icons.AutoMirrored.Filled.KeyboardArrowRight, contentDescription = null)
                        item.shortcut != null ->
                            Text(
                                text = item.shortcut,
                                color = KozmosThemeTokens.primitivesColorsForeground500
                            )
                    }
                }
            )
        }
    }
}

File: packages/ios/Sources/Components/Menu/Menu.swift
import SwiftUI

public enum KozmosMenuContentType: String, CaseIterable {
    case basic
    case checkbox
    case radio
    case submenu
}

public struct KozmosMenuItem: Identifiable, Hashable {
    public let id: String
    public let text: String
    public let shortcut: String?
    public let isChecked: Bool
    public let isSelected: Bool
    public let submenuItems: [String]

    public init(
        id: String? = nil,
        text: String,
        shortcut: String? = nil,
        isChecked: Bool = false,
        isSelected: Bool = false,
        submenuItems: [String] = []
    ) {
        self.id = id ?? text
        self.text = text
        self.shortcut = shortcut
        self.isChecked = isChecked
        self.isSelected = isSelected
        self.submenuItems = submenuItems
    }
}

public struct KozmosMenuContent: Hashable {
    public let label: String?
    public let items: [KozmosMenuItem]
    public let contentType: KozmosMenuContentType

    public init(
        label: String? = nil,
        items: [KozmosMenuItem],
        contentType: KozmosMenuContentType = .basic
    ) {
        self.label = label
        self.items = items
        self.contentType = contentType
    }
}

public struct KozmosMenu: View {
    let title: String
    let content: KozmosMenuContent
    let onItemSelect: (KozmosMenuItem) -> Void
    
    public init(title: String, items: [String], onSelect: @escaping (String) -> Void) {
        self.title = title
        self.content = KozmosMenuContent(items: items.map { KozmosMenuItem(text: $0) })
        self.onItemSelect = { item in onSelect(item.text) }
    }

    public init(
        title: String,
        content: KozmosMenuContent,
        onItemSelect: @escaping (KozmosMenuItem) -> Void = { _ in }
    ) {
        self.title = title
        self.content = content
        self.onItemSelect = onItemSelect
    }
    
    public var body: some View {
        Menu(title) {
            if let label = content.label {
                Text(label)
                    .font(KozmosTypography.caption)
                    .foregroundColor(KozmosColors.primitivesColorsForeground500)
                Divider().overlay(KozmosColors.semanticsBorderSubtle)
            }

            ForEach(content.items) { item in
                menuRow(for: item)
            }
        }
    }

    @ViewBuilder
    private func menuRow(for item: KozmosMenuItem) -> some View {
        switch content.contentType {
        case .basic:
            Button {
                onItemSelect(item)
            } label: {
                HStack {
                    Text(item.text)

                    if let shortcut = item.shortcut {
                        Spacer()
                        Text(shortcut)
                            .foregroundColor(KozmosColors.primitivesColorsForeground500)
                    }
                }
            }
        case .checkbox:
            Button {
                onItemSelect(item)
            } label: {
                Label(item.text, systemImage: item.isChecked ? "checkmark" : "square")
            }
        case .radio:
            Button {
                onItemSelect(item)
            } label: {
                Label(item.text, systemImage: item.isSelected ? "largecircle.fill.circle" : "circle")
            }
        case .submenu:
            if item.submenuItems.isEmpty {
                Button(item.text) {
                    onItemSelect(item)
                }
            } else {
                Menu(item.text) {
                    ForEach(item.submenuItems, id: \.self) { submenuItem in
                        Button(submenuItem) {
                            onItemSelect(
                                KozmosMenuItem(
                                    id: "\(item.id)/\(submenuItem)",
                                    text: submenuItem
                                )
                            )
                        }
                    }
                }
            }
        }
    }
}
